packages/backend-common/src/database/DatabaseManager.ts

```diff
@@ -95,10 +96,16 @@ export class DatabaseManager {
     const connection = this.getConnectionConfig(pluginId);
 
     if (this.getClientType(pluginId).client === 'sqlite3') {
+      const sqliteFilename = (connection as Knex.Sqlite3ConnectionConfig)
+        ?.filename;
+
+      // if persisting to a file, create separate files per plugin to avoid db migration issues.
+      if (sqliteFilename && sqliteFilename !== ':memory:') {
+        return path.join(sqliteFilename, `${pluginId}.sqlite`);
```

Comment:
I guess I have one remaining question.

You can configure the connection both globally, and per plugin. If you say _for a plugin_ that its `filename` parameter is something - I think it really sounds from the end user's point of view that they'd be typing in a path including the file name. It's just a bit odd to define a directory in a plugin config.

But at the same time, giving a full file name on the global level absolutely does _not_ make sense, as this PR neatly addresses.

I could see a few ways of align those differences:

- Make the `filename` always be a pattern that gets filled in. You can specify e.g. `'data/${{ plugin.id }}.sqlite'` which then uses the same mechanism / syntax as the new templates, or
- Make the `filename` be treated precisely like in this PR as a folder unconditionally, and for plugins that specify a filename, they are treated as files unconditionally.
- Like the one above but with additional checks that the folder _cannot_ end in `.sqlite` but the files _must_.

Any preferences there?

Reply:
Yeah thinking a bit more about this it's a bit confusing providing a filename and then realise it's actually used to construct a path. I would probably have named it something like `backstage.sqlite` and be confused to see that result in a folder with that name.

Reply:
> Add a completely custom new key instead called directory or folder, and when we find that one we insert a filename computed from it, and delete the custom one. That would maintain the filename semantic intact and perhaps be more clear.

Leaning toward this option because it has the least "surprising" behavior and builds on the current implementation. Should the `directory` config be global only or allowed per plugin? I think we'd still need to create a plugin folder within this directory, right? So with `directory=sqlite-dbs, filename=backstage.sqlite` the generated files would be in `sqlite-dbs/{pluginId}/backstage.sqlite`? Maybe this extra folder can be skipped if a plugin-specific `filename` is provided.

Reply:
Oh I actually meant (awkwardly typing on mobile) that there'd be no merging between the global and plugin levels. If you set `filename`, whether global or plugin level, then that's passed through verbatim to knex. Choosing to do so in the global one, will then lead to the old broken behavior, sure, so at the global level you would use `directory`. But setting `a.sqlite` on the plugin level leads to exactly `a.sqlite` no matter what the global setting was. I dunno, I hope that's fairly clear to the user, feels like all solutions have tradeoffs :) but I was assuming for leaving the builtin knex behavior completely untouched, and only adding the `directory` thing as a custom addon on the global level

Reply:
Yeah agreed on keeping this behavior out of the knex layer. Going back to the `directory=sqlite-dbs, filename=backstage.sqlite` example, does the `sqlite-dbs/{pluginId}/backstage.sqlite` output make sense? (assuming no plugin config overrides are provided)

Reply:
Aha you meant if _both_ were given in _one_ place (the global config). Sorry I misunderstood. Um, yeah probably that is less surprising than overwriting/ignoring the filename which I'd actually first thought (first spontaneous gut reaction was to essentially consider it a misconfiguration to specify both, but your idea is better)

Reply:
We could also prevent `filename` in the global config 🤔 

For example:

```
database:
  connection:
    directory: sqlite-dbs
  plugin:
    auth:
      connection:
        filename: custom-auth.sqlite
```

would produce

```
/sqlite-dbs/{pluginId}.sqlite # auth excluded
/custom-auth.sqlite
```

Reply:
Hm maybe. But I'm thinking that your first proposal is nice too, it's probably the case that people won't mind much the actual specifics of what the full path turns out to be. So as long as it acts sensibly and fixes the bug at the same time, you're probably golden.

Reply:
Essentially - doing something that's kinda fine and works, no matter what config you give, may trump something that's more unforgiving

Reply:
Just pushed changes to support `connection.directory` and some tests that hopefully explain the behavior. Considering a top level `filename` won't work due to migration conflicts, I decided to throw an error if that's provided with a message that suggests `directory` or a plugin specific `filename` instead.